### PR TITLE
Port remaining commands to use AbstractCommand

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Continuation of https://github.com/Homebrew/brew/pull/16975

Resolves https://github.com/Homebrew/brew/issues/16814 (though there will be some cleanup PRs)

### CODE REVIEW

File: Library/Homebrew/cmd/--prefix.rb [hunk 1]
```diff
@@ -1,106 +1,114 @@
...
-  end
-
-  def self.__prefix
-    args = __prefix_args.parse
+  module Cmd
+    class Prefix < AbstractCommand
+      include FileUtils
+
+      UNBREWED_EXCLUDE_FILES = %w[.DS_Store].freeze
+      UNBREWED_EXCLUDE_PATHS = %w[
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/--prefix.rb:13]: Would it be too paranoid to make these private constants?

File: Library/Homebrew/cmd/outdated.rb [hunk 2]
```diff
@@ -1,200 +1,202 @@
...
+        end
       end
-    end
-  end
 
-  def self.verbose?
-    ($stdout.tty? || super) && !quiet?
-  end
+      def verbose?
+        ($stdout.tty? || Context.current.verbose?) && !Context.current.quiet?
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/outdated.rb:152]: The context mixin is another option here.

File: Library/Homebrew/cmd/reinstall.rb [hunk 3]
```diff
@@ -15,172 +15,174 @@
...
+
+        Install.perform_preinstall_checks
+
+        formulae.each do |formula|
+          if formula.pinned?
+            onoe "#{formula.full_name} is pinned. You must unpin it to reinstall."
+            next
+          end
+          Migrator.migrate_if_needed(formula, force: args.force?)
+          Homebrew.reinstall_formula(
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/reinstall.rb:136]: This should probably get namespaced properly at some point. It seems to get used in both the upgrade and reinstall commands.

File: Library/Homebrew/test/cmd/rbenv-sync_spec.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+require "cmd/rbenv-sync"
+require "cmd/shared_examples/args_parse"
+
+RSpec.describe Homebrew::Cmd::RbenvSync do
+  it_behaves_like "parseable arguments"
```
Comment (apainintheneck) [inline Library/Homebrew/test/cmd/rbenv-sync_spec.rb:7]: I wonder if we could just automatically run `it_behaves_like "parseable arguments"` on all abstract command classes. The would remove the need to add this for every command and make it impossible to forget to add this test.

File: Library/Homebrew/cmd/untap.rb
Comment (apainintheneck) [inline Library/Homebrew/cmd/untap.rb:1]: I'll move the untap module back into this class after this gets merged in as discussed in https://github.com/Homebrew/brew/issues/16814#issuecomment-1996318943.

**Review by apainintheneck — APPROVED** [review]
You're a machine. Are you using the dark arts (sed) to update these automatically? 🤣

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @dduugg and to @apainintheneck for the review!

**Comment by mahmudzero** [comment]
This seems to have broken some commands on some linux machines, running into a
```
Error: cannot load such file -- abstract_command
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-bundle/cmd/bundle.rb:3:in `require'
```

When attempting a `brew bundle --file=...`, will take a look to see if there is anything I can catch but I'm not familiar with the brew code base so I'm sure by the time I catch it one of you will have a fix... lol

**Comment by mahmudzero** [comment]
> > after a
> > ```
> > $ brew update
> > ```
> 
> Please do not complain about things breaking before running `brew update`.

Thats a fair comment, that said, for some reason the install script (https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh), in certain environments, is installing an outdated version of brew - I will take a look into it.

Also wasn't really complaining and I closed my own comment as off topic ¯\_(ツ)_/¯, really appreciate the work y'all do was just jarring this morning (I have a GitHub Action that installs brew each time, using the install script, and for some reason it's installing 4.2.14, instead of the latest version, and this was a change that happened over night but I see no updates to the install script in a few weeks so IDK what's going on there and not blaming anyone anyway)

**Pull request merged**